File: packages/flutter/ios/Classes/AuraVerifierPlugin.swift
import Flutter
import UIKit

/// Flutter plugin for Aura Verifier SDK
public class AuraVerifierPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "aura_verifier",
            binaryMessenger: registrar.messenger()
        )
        let instance = AuraVerifierPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "verifyCredential":
            handleVerifyCredential(call, result: result)

        case "parseQRCode":
            handleParseQRCode(call, result: result)

        case "checkCredentialStatus":
            handleCheckCredentialStatus(call, result: result)

        case "getCachedCredentials":
            handleGetCachedCredentials(call, result: result)

        case "clearCache":
            handleClearCache(call, result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleVerifyCredential(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let qrCodeData = args["qrCodeData"] as? String,
              let restEndpoint = args["restEndpoint"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments",
                details: nil
            ))
            return
        }

        let verifierAddress = args["verifierAddress"] as? String
        let timeoutMs = args["timeoutMs"] as? Int ?? 10000

        // Perform verification via REST API
        verifyViaREST(
            qrCodeData: qrCodeData,
            restEndpoint: restEndpoint,
            verifierAddress: verifierAddress,
            timeoutMs: timeoutMs
        ) { verificationResult, error in
            if let error = error {
                result(FlutterError(
                    code: "VERIFICATION_ERROR",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(verificationResult)
            }
        }
    }

    private func handleParseQRCode(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let qrCodeData = args["qrCodeData"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing qrCodeData",
                details: nil
            ))
            return
        }

        do {
            let parsed = try parseQRCodeData(qrCodeData)
            result(parsed)
        } catch {
            result(FlutterError(
                code: "PARSE_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    private func handleCheckCredentialStatus(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let vcId = args["vcId"] as? String,
              let restEndpoint = args["restEndpoint"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments",
                details: nil
            ))
            return
        }

        checkVCStatus(
            vcId: vcId,
            restEndpoint: restEndpoint
        ) { status, error in
            if let error = error {
                result(FlutterError(
                    code: "STATUS_CHECK_ERROR",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(status)
            }
        }
    }

    private func handleGetCachedCredentials(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // In a production implementation, this would retrieve cached credentials
        // from secure storage (Keychain on iOS)
        result([])
    }

    private func handleClearCache(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // In a production implementation, this would clear the Keychain
        result(true)
    }

    // MARK: - Helper Methods

    private func verifyViaREST(
        qrCodeData: String,
        restEndpoint: String,
        verifierAddress: String?,
        timeoutMs: Int,
        completion: @escaping ([String: Any]?, Error?) -> Void
    ) {
        let urlString = "\(restEndpoint)/aura/vcregistry/v1beta1/verify_presentation"
        guard let url = URL(string: urlString) else {
            completion(nil, NSError(
                domain: "AuraVerifier",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid URL"]
            ))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = TimeInterval(timeoutMs) / 1000.0

        var body: [String: Any] = ["qr_code_data": qrCodeData]
        if let verifierAddress = verifierAddress {
            body["verifier_address"] = verifierAddress
        }

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            completion(nil, error)
            return
        }

        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(nil, error)
                return
            }

            guard let data = data,
                  let httpResponse = response as? HTTPURLResponse else {
                completion(nil, NSError(
                    domain: "AuraVerifier",
                    code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid response"]
                ))
                return
            }

            if httpResponse.statusCode == 200 {
                do {
                    if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                        let result = json["result"] as? [String: Any] ?? json
                        completion(result, nil)
                    } else {
                        completion(nil, NSError(
                            domain: "AuraVerifier",
                            code: -1,
                            userInfo: [NSLocalizedDescriptionKey: "Invalid JSON response"]
                        ))
                    }
                } catch {
                    completion(nil, error)
                }
            } else {
                completion(nil, NSError(
                    domain: "AuraVerifier",
                    code: httpResponse.statusCode,
                    userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"]
                ))
            }
        }

        task.resume()
    }

    private func checkVCStatus(
        vcId: String,
        restEndpoint: String,
        completion: @escaping ([String: Any]?, Error?) -> Void
    ) {
        let urlString = "\(restEndpoint)/aura/vcregistry/v1beta1/vc_status/\(vcId)"
        guard let url = URL(string: urlString) else {
            completion(nil, NSError(
                domain: "AuraVerifier",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid URL"]
            ))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(nil, error)
                return
            }

            guard let data = data,
                  let httpResponse = response as? HTTPURLResponse else {
                completion(nil, NSError(
                    domain: "AuraVerifier",
                    code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid response"]
                ))
                return
            }

            if httpResponse.statusCode == 200 {
                do {
                    if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                        completion(json, nil)
                    } else {
                        completion(nil, NSError(
                            domain: "AuraVerifier",
                            code: -1,
                            userInfo: [NSLocalizedDescriptionKey: "Invalid JSON response"]
                        ))
                    }
                } catch {
                    completion(nil, error)
                }
            } else {
                completion(nil, NSError(
                    domain: "AuraVerifier",
                    code: httpResponse.statusCode,
                    userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"]
                ))
            }
        }

        task.resume()
    }

    private func parseQRCodeData(_ qrCodeData: String) throws -> [String: Any] {
        var base64Data = qrCodeData

        // Remove URL prefix if present
        if qrCodeData.hasPrefix("aura://verify?data=") {
            base64Data = String(qrCodeData.dropFirst("aura://verify?data=".count))
        }

        // Decode base64
        guard let data = Data(base64Encoded: base64Data),
              let jsonString = String(data: data, encoding: .utf8),
              let jsonData = jsonString.data(using: .utf8),
              let json = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
            throw NSError(
                domain: "AuraVerifier",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid QR code format"]
            )
        }

        return json
    }
}

File: packages/flutter/android/src/main/kotlin/com/aura/verifier/AuraVerifierPlugin.kt
package com.aura.verifier

import android.util.Base64
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/** AuraVerifierPlugin */
class AuraVerifierPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "aura_verifier")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "verifyCredential" -> {
                handleVerifyCredential(call, result)
            }

            "parseQRCode" -> {
                handleParseQRCode(call, result)
            }

            "checkCredentialStatus" -> {
                handleCheckCredentialStatus(call, result)
            }

            "getCachedCredentials" -> {
                handleGetCachedCredentials(call, result)
            }

            "clearCache" -> {
                handleClearCache(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // MARK: - Method Handlers

    private fun handleVerifyCredential(call: MethodCall, result: Result) {
        val qrCodeData = call.argument<String>("qrCodeData")
        val restEndpoint = call.argument<String>("restEndpoint")
        val verifierAddress = call.argument<String?>("verifierAddress")
        val timeoutMs = call.argument<Int>("timeoutMs") ?: 10000

        if (qrCodeData == null || restEndpoint == null) {
            result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
            return
        }

        scope.launch {
            try {
                val verificationResult = verifyViaREST(
                    qrCodeData,
                    restEndpoint,
                    verifierAddress,
                    timeoutMs
                )
                result.success(verificationResult)
            } catch (e: Exception) {
                result.error("VERIFICATION_ERROR", e.message, null)
            }
        }
    }

    private fun handleParseQRCode(call: MethodCall, result: Result) {
        val qrCodeData = call.argument<String>("qrCodeData")

        if (qrCodeData == null) {
            result.error("INVALID_ARGUMENTS", "Missing qrCodeData", null)
            return
        }

        try {
            val parsed = parseQRCodeData(qrCodeData)
            result.success(parsed)
        } catch (e: Exception) {
            result.error("PARSE_ERROR", e.message, null)
        }
    }

    private fun handleCheckCredentialStatus(call: MethodCall, result: Result) {
        val vcId = call.argument<String>("vcId")
        val restEndpoint = call.argument<String>("restEndpoint")

        if (vcId == null || restEndpoint == null) {
            result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
            return
        }

        scope.launch {
            try {
                val status = checkVCStatus(vcId, restEndpoint)
                result.success(status)
            } catch (e: Exception) {
                result.error("STATUS_CHECK_ERROR", e.message, null)
            }
        }
    }

    private fun handleGetCachedCredentials(call: MethodCall, result: Result) {
        // In a production implementation, this would retrieve cached credentials
        // from secure storage (EncryptedSharedPreferences on Android)
        result.success(emptyList<Map<String, Any>>())
    }

    private fun handleClearCache(call: MethodCall, result: Result) {
        // In a production implementation, this would clear the secure storage
        result.success(true)
    }

    // MARK: - Helper Methods

    private suspend fun verifyViaREST(
        qrCodeData: String,
        restEndpoint: String,
        verifierAddress: String?,
        timeoutMs: Int
    ): Map<String, Any> = withContext(Dispatchers.IO) {
        val urlString = "$restEndpoint/aura/vcregistry/v1beta1/verify_presentation"
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection

        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("Accept", "application/json")
            connection.connectTimeout = timeoutMs
            connection.readTimeout = timeoutMs
            connection.doOutput = true

            // Build request body
            val body = JSONObject().apply {
                put("qr_code_data", qrCodeData)
                verifierAddress?.let { put("verifier_address", it) }
            }

            // Write request body
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(body.toString())
                writer.flush()
            }

            // Read response
            if (connection.responseCode == HttpURLConnection.HTTP_OK) {
                val response = BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                    reader.readText()
                }

                val json = JSONObject(response)
                val result = if (json.has("result")) {
                    json.getJSONObject("result")
                } else {
                    json
                }

                return@withContext jsonToMap(result)
            } else {
                throw Exception("HTTP ${connection.responseCode}")
            }
        } finally {
            connection.disconnect()
        }
    }

    private suspend fun checkVCStatus(
        vcId: String,
        restEndpoint: String
    ): Map<String, Any> = withContext(Dispatchers.IO) {
        val urlString = "$restEndpoint/aura/vcregistry/v1beta1/vc_status/$vcId"
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection

        try {
            connection.requestMethod = "GET"
            connection.setRequestProperty("Accept", "application/json")

            if (connection.responseCode == HttpURLConnection.HTTP_OK) {
                val response = BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                    reader.readText()
                }

                val json = JSONObject(response)
                return@withContext jsonToMap(json)
            } else {
                throw Exception("HTTP ${connection.responseCode}")
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun parseQRCodeData(qrCodeData: String): Map<String, Any> {
        var base64Data = qrCodeData

        // Remove URL prefix if present
        if (qrCodeData.startsWith("aura://verify?data=")) {
            base64Data = qrCodeData.substring("aura://verify?data=".length)
        }

        // Decode base64
        val decoded = Base64.decode(base64Data, Base64.DEFAULT)
        val jsonString = String(decoded, Charsets.UTF_8)
        val json = JSONObject(jsonString)

        return jsonToMap(json)
    }

    private fun jsonToMap(json: JSONObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        json.keys().forEach { key ->
            val value = json.get(key)
            map[key] = when (value) {
                is JSONObject -> jsonToMap(value)
                JSONObject.NULL -> null as Any
                else -> value
            }
        }
        return map
    }
}
